File: packages/tx402/test/conformance/handlers.ts
/**
 * Stage B handlers — the TypeScript SDK executed against the shared vectors.
 *
 * One handler per vector `kind`. Registering a handler is what claims the kind; the runner
 * fails if a vector at or below {@link IMPLEMENTED_THROUGH} has none, so this file and
 * `IMPLEMENTED_THROUGH` move together.
 *
 * Handlers throw on mismatch rather than returning false, because the diff is the only
 * genuinely useful part of a conformance failure.
 */

import { expect } from "vitest";

import { canonicalizeJson, CanonicalJsonError } from "../../src/core/canonical-json.js";
import { TX402_ERROR_TAXONOMY } from "../../src/core/errors.js";
import { isTx402Error } from "../../src/core/errors.js";
import {
  resolveNetwork,
  verifyReleaseManifest,
  type ReleaseManifest,
} from "../../src/core/manifest.js";
import { decodePaymentRequired } from "../../src/core/protocol.js";
import { registerHandler, type ConformanceVector } from "./runner.js";

/** Manifest failures all surface to callers as ConfigurationError (SPEC §5.4). */
const MANIFEST_ERROR_CODE = "TX402_CONFIG_INVALID";

registerHandler("errors.taxonomy", (vector: ConformanceVector) => {
  const expected = vector.expected as {
    entries: {
      code: string;
      className: string;
      retryability: string;
      retryable: boolean;
      requiredDetails: string[];
    }[];
  };

  // Compared as whole arrays, in order: the taxonomy's ordering is part of what is frozen,
  // and an entry-by-entry loop would let a reordering pass.
  const actual = TX402_ERROR_TAXONOMY.map((entry) => ({
    code: entry.code,
    className: entry.className,
    retryability: entry.retryability,
    retryable: entry.retryable,
    requiredDetails: [...entry.requiredDetails],
  }));

  expect(actual).toEqual(expected.entries);
});

registerHandler("canonical-json", (vector: ConformanceVector) => {
  const { document } = vector.input as { document: unknown };
  const expected = vector.expected as
    { canonical: string; sha256: string } | { error: string };

  if ("error" in expected) {
    try {
      canonicalizeJson(document);
    } catch (error) {
      expect(error).toBeInstanceOf(CanonicalJsonError);
      expect((error as CanonicalJsonError).reason).toBe(expected.error);
      return;
    }
    throw new Error(
      `Expected canonicalization to fail with ${expected.error}, but it succeeded`,
    );
  }

  expect(canonicalizeJson(document)).toBe(expected.canonical);
});

registerHandler("manifest.verify", (vector: ConformanceVector) => {
  const input = vector.input as {
    manifest: unknown;
    nowEpochMs: number;
    trustedKeys?: Record<string, string>;
  };
  const expected = vector.expected as
    { outcome: "valid" } | { outcome: "invalid"; errorCode: string; reason: string };

  const result = verifyReleaseManifest(input.manifest, {
    nowEpochMs: input.nowEpochMs,
    ...(input.trustedKeys ? { trustedKeys: input.trustedKeys } : {}),
  });

  if (expected.outcome === "valid") {
    if (!result.valid) {
      throw new Error(
        `Expected the manifest to verify, but it failed: ${result.reason} — ${result.message}`,
      );
    }
    return;
  }

  if (result.valid) {
    throw new Error(
      `Expected the manifest to be rejected with ${expected.reason}, but it verified`,
    );
  }

  expect(result.reason).toBe(expected.reason);
  expect(MANIFEST_ERROR_CODE).toBe(expected.errorCode);
});

registerHandler("manifest.network-resolution", (vector: ConformanceVector) => {
  const input = vector.input as { manifest: ReleaseManifest; query: string };
  const expected = vector.expected as
    { resolved: string; wasAlias: boolean } | { errorCode: string; reason: string };

  const result = resolveNetwork(input.manifest, input.query);

  if ("resolved" in expected) {
    if (!("resolved" in result)) {
      throw new Error(
        `Expected ${input.query} to resolve to ${expected.resolved}, but it failed: ${result.message}`,
      );
    }
    expect(result.resolved).toBe(expected.resolved);
    expect(result.wasAlias).toBe(expected.wasAlias);
    return;
  }

  if ("resolved" in result) {
    throw new Error(
      `Expected ${input.query} to be rejected, but it resolved to ${result.resolved}`,
    );
  }
  expect(result.reason).toBe(expected.reason);
  expect(MANIFEST_ERROR_CODE).toBe(expected.errorCode);
});

registerHandler("protocol.decode-payment-required", (vector: ConformanceVector) => {
  const input = vector.input as {
    requestUrl: string;
    requestMethod: string;
    header?: string;
    generatedHeader?: { kind: "repeated-ascii"; decodedBytes: number };
    clockEpochMs: number;
  };
  const expected = vector.expected as
    | { outcome: "valid"; normalized: unknown }
    | { outcome: "invalid"; errorCode: string; reason: string };

  try {
    const header = input.generatedHeader
      ? Buffer.alloc(input.generatedHeader.decodedBytes, 0x78).toString("base64")
      : input.header;
    const normalized = decodePaymentRequired(header, {
      requestUrl: input.requestUrl,
      requestMethod: input.requestMethod,
      requestId: vector.id,
      clockEpochMs: input.clockEpochMs,
    });
    if (expected.outcome === "invalid") {
      throw new Error(`Expected decode to fail with ${expected.reason}, but it succeeded`);
    }
    expect(normalized).toEqual(expected.normalized);
  } catch (error) {
    if (expected.outcome === "valid") throw error;
    if (!isTx402Error(error)) throw error;
    expect(error.code).toBe(expected.errorCode);
    expect(error.details.reason).toBe(expected.reason);
  }
});

File: packages/tx402-python/tests/conformance/handlers.py
"""Stage B handlers — the Python SDK executed against the shared vectors.

One handler per vector ``kind``. Registering a handler is what claims the kind; the runner
fails if a vector at or below :data:`IMPLEMENTED_THROUGH` has none, so this module and that
constant move together.

Handlers raise on mismatch rather than returning ``False``, because the diff is the only
genuinely useful part of a conformance failure.
"""

from __future__ import annotations

import base64
import hashlib
from typing import Any, Final

import pytest

from tests.conformance.runner import register_handler
from tx402.canonical_json import CanonicalJsonError, canonicalize_json
from tx402.errors import TX402_ERROR_TAXONOMY, Tx402Error
from tx402.fingerprint import (
    digest_request_body,
    fingerprint_request,
    normalize_fingerprint_url,
)
from tx402.ledger import MemorySpendStore
from tx402.manifest import resolve_network, verify_release_manifest
from tx402.protocol import decode_payment_required

#: Manifest failures all surface to callers as ConfigurationError (SPEC §5.4).
MANIFEST_ERROR_CODE: Final = "TX402_CONFIG_INVALID"


def _errors_taxonomy(vector: dict[str, Any]) -> None:
    expected = vector["expected"]["entries"]

    # Compared as whole lists, in order: the taxonomy's ordering is part of what is frozen,
    # and an entry-by-entry loop would let a reordering pass.
    actual = [
        {
            "code": entry.code,
            "className": entry.class_name,
            "retryability": entry.retryability,
            "retryable": entry.retryable,
            "requiredDetails": list(entry.required_details),
        }
        for entry in TX402_ERROR_TAXONOMY
    ]

    assert actual == expected


def _canonical_json(vector: dict[str, Any]) -> None:
    document = vector["input"]["document"]
    expected = vector["expected"]

    if "error" in expected:
        with pytest.raises(CanonicalJsonError) as raised:
            canonicalize_json(document)
        # The reason, not merely the failure: two implementations that reject the same
        # input for different reasons have not agreed on anything useful.
        assert raised.value.reason == expected["error"]
        return

    canonical = canonicalize_json(document)
    assert canonical == expected["canonical"]

    digest = f"sha256:{hashlib.sha256(canonical.encode('ascii')).hexdigest()}"
    assert digest == expected["sha256"]


def _manifest_verify(vector: dict[str, Any]) -> None:
    payload = vector["input"]
    expected = vector["expected"]
    result = verify_release_manifest(
        payload["manifest"],
        now_epoch_ms=payload["nowEpochMs"],
        trusted_keys=payload.get("trustedKeys"),
    )

    if expected["outcome"] == "valid":
        assert result.valid, (
            f"Expected the manifest to verify, but it failed: "
            f"{result.reason} — {result.message}"
        )
        return

    assert not result.valid, (
        f"Expected the manifest to be rejected with {expected['reason']}, but it verified"
    )
    assert result.reason == expected["reason"]
    assert expected["errorCode"] == MANIFEST_ERROR_CODE


def _manifest_network_resolution(vector: dict[str, Any]) -> None:
    payload = vector["input"]
    expected = vector["expected"]

    result = resolve_network(payload["manifest"], payload["query"])

    if "resolved" in expected:
        assert result.resolved is not None, (
            f"Expected {payload['query']} to resolve to {expected['resolved']}, "
            f"but it failed: {result.message}"
        )
        assert result.resolved == expected["resolved"]
        assert result.was_alias == expected["wasAlias"]
        return

    assert result.resolved is None, (
        f"Expected {payload['query']} to be rejected, but it resolved to {result.resolved}"
    )
    assert result.reason == expected["reason"]
    assert expected["errorCode"] == MANIFEST_ERROR_CODE


register_handler("errors.taxonomy", _errors_taxonomy)
register_handler("canonical-json", _canonical_json)
register_handler("manifest.verify", _manifest_verify)
register_handler("manifest.network-resolution", _manifest_network_resolution)


def _protocol_decode_payment_required(vector: dict[str, Any]) -> None:
    payload = vector["input"]
    expected = vector["expected"]
    header = payload.get("header")
    if "generatedHeader" in payload:
        header = base64.b64encode(
            b"x" * int(payload["generatedHeader"]["decodedBytes"])
        ).decode("ascii")
    arguments: dict[str, Any] = {
        "request_url": payload["requestUrl"],
        "request_method": payload["requestMethod"],
        "request_id": vector["id"],
        "clock_epoch_ms": payload["clockEpochMs"],
    }
    if expected["outcome"] == "invalid":
        with pytest.raises(Tx402Error) as raised:
            decode_payment_required(header, **arguments)
        assert raised.value.code == expected["errorCode"]
        assert raised.value.details.get("reason") == expected["reason"]
        return

    normalized = decode_payment_required(
        header,
        **arguments,
    )

    assert expected["outcome"] == "valid"
    assert normalized == expected["normalized"]


register_handler("protocol.decode-payment-required", _protocol_decode_payment_required)


def _request_fingerprint(vector: dict[str, Any]) -> None:
    payload = vector["input"]
    expected = vector["expected"]
    assert normalize_fingerprint_url(payload["url"]) == expected["normalizedUrl"]
    assert digest_request_body(payload["body"]) == expected["bodyHash"]
    assert (
        fingerprint_request(
            method=payload["method"],
            url=payload["url"],
            body=payload["body"],
            challenge_hash=payload["challengeHash"],
        )
        == expected["fingerprint"]
    )


def _spend_ledger_behavior(vector: dict[str, Any]) -> None:
    store = MemorySpendStore()
    outcomes: list[dict[str, Any]] = []
    for operation in vector["input"]["operations"]:
        action = operation["action"]
        try:
            if action == "reserve":
                reservation = store.reserve(
                    reservation_id=operation["reservationId"],
                    request_id=operation["requestId"],
                    policy_scope=operation["policyScope"],
                    request_fingerprint=operation["requestFingerprint"],
                    asset_id=operation["assetId"],
                    amount_atomic=operation["amountAtomic"],
                    max_per_hour_atomic=operation["maxPerHourAtomic"],
                    now_epoch_ms=operation["nowEpochMs"],
                )
                outcomes.append({"outcome": "reserved", "state": reservation.state})
            elif action == "commit":
                store.commit(
                    reservation_id=operation["reservationId"],
                    committed_at_epoch_ms=operation["committedAtEpochMs"],
                    settlement_id=operation.get("settlementId"),
                )
                outcomes.append({"outcome": "committed"})
            elif action == "release":
                reservation = store.release(
                    reservation_id=operation["reservationId"],
                    now_epoch_ms=operation["nowEpochMs"],
                )
                outcomes.append({"outcome": "released", "state": reservation.state})
            elif action == "snapshot":
                state = store.get_budget_state(
                    policy_scope=operation["policyScope"],
                    asset_id=operation["assetId"],
                    now_epoch_ms=operation["nowEpochMs"],
                )
                outcomes.append(
                    {
                        "outcome": "snapshot",
                        "committedAtomic": state.committed_atomic,
                        "reservedAtomic": state.reserved_atomic,
                        "reservationStates": [
                            reservation.state for reservation in state.reservations
                        ],
                        "entryCount": len(state.entries),
                    }
                )
            else:
                raise ValueError(f"Unknown ledger operation {action}")
        except Tx402Error as error:
            outcomes.append({"outcome": "error", "errorCode": error.code})
    assert outcomes == vector["expected"]["outcomes"]


register_handler("request.fingerprint", _request_fingerprint)
register_handler("spend-ledger.behavior", _spend_ledger_behavior)
